# 01. Importing Libraries

In [40]:
# Import libraries
import pandas as pd
import numpy as np
import os


# 02 Import Data

In [41]:
path = r'C:\Users\tee3a\3D Objects\12-2022 Instacart Basket Analysis'

# 03 products.csv” data set from your “Original Data

In [42]:
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data for raw data sets', 'products.csv'), index_col = False)

# 04 orders_wrangled.csv” data set from your “Prepared Data” folder as df_ords

In [43]:
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data for storing any manipulated data sets', 'orders_wrangled.csv'), index_col = False)

# 05 new test dataframe

In [44]:
df_test = pd.DataFrame()

In [45]:
df_test['mix'] = ['a', 'b', 1, True]

In [46]:
df_test['mix']

0       a
1       b
2       1
3    True
Name: mix, dtype: object

In [47]:
df_test.head()

mix
0     a
1     b
2     1
3  True

# 06 fixing mixed-type column(numeric-int64, string-str)

In [48]:
df_test['mix'] = df_test['mix'].astype('str')

In [49]:
df_test['mix']

0       a
1       b
2       1
3    True
Name: mix, dtype: object

# 07 Finding Missing Values

In [50]:
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

# 08 identifying the rows with missing values


In [51]:
df_nan =  df_prods[df_prods['product_name'].isnull() == True]

In [52]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

# 09 Addressing Missing Values, code to replace your missing values

In [53]:
df_prods.shape

(49693, 5)

# 10 create a new dataframe to clean products


In [54]:
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [55]:
df_prods_clean.shape

(49677, 5)

# 11 Finding Duplicates

In [56]:
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [57]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [58]:
df_prods_clean.shape

(49677, 5)

In [59]:
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [60]:
df_prods_clean_no_dups.shape

(49672, 5)

In [61]:
df_prods.shape

(49693, 5)

In [62]:
df_prods.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

In [63]:
df_prods_clean_no_dups.describe()

product_id      aisle_id  department_id        prices
count  49672.000000  49672.000000   49672.000000  49672.000000
mean   24850.349775     67.762442      11.728942      9.993282
std    14340.705287     38.315784       5.850779    453.615536
min        1.000000      1.000000       1.000000      1.000000
25%    12432.750000     35.000000       7.000000      4.100000
50%    24850.500000     69.000000      13.000000      7.100000
75%    37268.250000    100.000000      17.000000     11.100000
max    49688.000000    134.000000      21.000000  99999.000000

 From the above, it is noted that the product_id, aisle_id and department_id are seen as numeric instead of string. Also, there is an outlier noticed in the prices column. The minmum value is 1, the mean is 9.9993282 and the maxiumum value is 99,999. So, the outlier is in the maximum.

# Task 4.5

In [64]:
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number     order_dow  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01  2.776219e+00   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01  2.046829e+00   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00  0.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00  1.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01  3.000000e+00   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01  5.000000e+00   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02  6.000000e+00   

       order_hour_of_day  days_since_prior_order  
count       3.421083e+06            3.214874e+06  
mean        1.345202e+01            1.111484e+01  
std         4.226088e+00            9.206737e+00  
min         0.000000e+00            0.000000e+00  
25%         1.000000e+01            4.000000e+00  
50%         1.300000e+01            7.000000e+00  
75%         1.600000e+01            1.500000e+01  
max         2.300000e+01            3.000000e+01

A new column 'Unnamed' just popped up, it needs to be investigated. There are no outliers or erroneous values noticed.

3. Check for mixed-type data in your df_ords dataframe.
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)
for col in df_ords.columns.tolist():

There is no mixed-type data in df_ords dataframe.

# 5. Run a check for missing values in your df_ords dataframe

In [65]:
df_ords.shape

(3421083, 8)

In [66]:
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
eval_set                       0
order_number                   0
order_dow                      0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

In [67]:
df_nan_2 =  df_ords[df_ords['days_since_prior_order'].isnull() == True]

In [68]:
df_nan_2

Unnamed: 0  order_id  user_id eval_set  order_number  order_dow  \
0                 0   2539329        1    prior             1          2   
11               11   2168274        2    prior             1          2   
26               26   1374495        3    prior             1          1   
39               39   3343014        4    prior             1          6   
45               45   2717275        5    prior             1          3   
...             ...       ...      ...      ...           ...        ...   
3420930     3420930    969311   206205    prior             1          4   
3420934     3420934   3189322   206206    prior             1          3   
3421002     3421002   2166133   206207    prior             1          6   
3421019     3421019   2227043   206208    prior             1          1   
3421069     3421069   3154581   206209    prior             1          3   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
11                      11                     NaN  
26                      14                     NaN  
39                      11                     NaN  
45                      12                     NaN  
...                    ...                     ...  
3420930                 12                     NaN  
3420934                 18                     NaN  
3421002                 19                     NaN  
3421019                 15                     NaN  
3421069                 11                     NaN  

[206209 rows x 8 columns]

days_since_prior_order has 206,209 missing values.The data might not have been recorded for the missing items.

# 6. Address the missing values using an appropriate method.

In [69]:
df_ords['days_since_prior_order'].fillna(1.111484, inplace=True)

In [70]:
df_nan_3 =  df_ords[df_ords['days_since_prior_order'].isnull() == True]

In [71]:
df_nan_3

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, eval_set, order_number, order_dow, order_hour_of_day, days_since_prior_order]
Index: []

The coumn with the missing values has a maxium number of 3 and a minimum of 1, there is no outliers in the data, ehnance I decided to replace the missing data with the mean of 1.111484

# 7. Run a check for duplicate values in your df_ords data.

In [72]:
df_ords_clean = df_ords[df_ords.isnull() == False]

In [73]:
df_dups_2 = df_ords_clean[df_ords_clean.duplicated()]


In [74]:
df_dups_2

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, eval_set, order_number, order_dow, order_hour_of_day, days_since_prior_order]
Index: []

There is no duplicates in df_ords data

# Drop the unwanted column

In [75]:
df_ords_1 = df_ords.drop(columns = ['Unnamed: 0'])

In [76]:
df_ords_1.shape

(3421083, 7)

# 8. Export Data

In [77]:
df_prods.to_csv(os.path.join(path, '02 Data','Prepared Data for storing any manipulated data sets', 'products_checked.csv'))

In [78]:
df_ords_1.to_csv(os.path.join(path, '02 Data','Prepared Data for storing any manipulated data sets', 'orders_checked.csv'))